### PR TITLE
Enable OnException for AspNetCore instrumentation

### LINKED ISSUE #1405: Exception recording in AspNetCore instrumentation
For AspNetCore instrumentation:
We have the ability to get exception from the diagnosticsource. We can add an ability to enrich the activity with this exception. Semantic conventions are here for exception: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md

https://github.com/dotnet/aspnetcore/blob/3fdbc2e90d9762324ef3809098c9ca25b0cb3284/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs#L141

https://github.com/dotnet/aspnetcore/blob/3892b4ea82b64aa7abe0a87c4277449c26b47c9c/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L231

### PR DESCRIPTION
**Pull request by eddynaka**
Fixes #1405.

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 1]
```diff
@@ -225,6 +226,29 @@ public override void OnCustom(string name, Activity activity, object payload)
...
+                try
+                {
+                    this.options.Enrich?.Invoke(activity, "OnException", exc);
+                }
+                catch (Exception ex)
+                {
+                    AspNetCoreInstrumentationEventSource.Log.EnrichmentException(ex);
+                }
+
+                activity.SetStatus(Status.Error.WithDescription(exc.Message));
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:248]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md Lets populate this into ActivityEvent.

**Comment by reyang** [comment]
@pakrym FYI.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs (hunk 1 above)
Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:249]: just updated and added in another place that was missing

**Comment by pakrym** [comment]
>@pakrym FYI.

Does this mean that Azure SDKs should put the message into description as well?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 2]
```diff
@@ -225,6 +226,30 @@ public override void OnCustom(string name, Activity activity, object payload)
...
+                {
+                    this.options.Enrich?.Invoke(activity, "OnException", exc);
+                }
+                catch (Exception ex)
+                {
+                    AspNetCoreInstrumentationEventSource.Log.EnrichmentException(ex);
+                }
+
+                activity.RecordException(exc);
+                activity.SetStatus(Status.Error.WithDescription(exc.Message));
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:249]: I think we should set Status with description, irrespective of `options.RecordException` flag.

**Comment by cijothomas** [comment]
> > @pakrym FYI.
> 
> Does this mean that Azure SDKs should put the message into description as well?

Not sure I understood. Could your clarify?
This PR is responding to one more DS event from AspNetCore. This event gives `Exception`, which is stored as `ActivityEvent` following the semantic conventions. There shouldn't be anything specific to Azure SDK here..

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs [hunk 3]
```diff
@@ -53,5 +53,10 @@ public class AspNetCoreInstrumentationOptions
         /// The type of this object depends on the event, which is given by the above parameter.</para>
         /// </remarks>
         public Action<Activity, string, object> Enrich { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the exception will be recorded or not.
+        /// </summary>
+        public bool RecordException { get; set; }
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:60]: Any thoughts on making this false by default? 
My suggestion is make this opt-in. Storing exception is somewhat expensive, so lets do this only if users opt-in

Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:60]: since we are not setting anything, it is false by default.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs (hunk 2 above)
Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:253]: done

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs [hunk 4]
```diff
@@ -184,6 +184,8 @@ public override void OnException(Activity activity, object payload)
                     HttpInstrumentationEventSource.Log.EnrichmentException(ex);
                 }
 
+                activity.RecordException(exc);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:187]: lets discuss and do this in separate PR.

**Comment by pakrym** [comment]
Recently @reyang mentioned in the email "We SHOULD NOT add any exception object to the Span. How to report an exception is not in the scope this year. I think most likely we will be using logs" but it looks like this PR uses `Status.Description` to report the exception message. So I'm trying to understand what should we do when reporting exceptions from Azure SDKs.

**Comment by cijothomas** [comment]
> Recently @reyang mentioned in the email "We SHOULD NOT add any exception object to the Span. How to report an exception is not in the scope this year. I think most likely we will be using logs" but it looks like this PR uses `Status.Description` to report the exception message. So I'm trying to understand what should we do when reporting exceptions from Azure SDKs.


> How to report an exception is not in the scope this year
This is probably specific to Microsoft specific Exporters. We can discuss it offline.

This PR is just making this repo compliant to the OpenTelemetry specs: 
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#status

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs (hunk 4 above)
Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:187]: ok. let me revert then

**Comment by CodeBlanch** [comment]
@eddynaka Are you going to add a test for this?

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs [hunk 5]
```diff
@@ -225,6 +226,34 @@ public override void OnCustom(string name, Activity activity, object payload)
...
+                try
+                {
+                    this.options.Enrich?.Invoke(activity, "OnException", exc);
+                }
+                catch (Exception ex)
+                {
+                    AspNetCoreInstrumentationEventSource.Log.EnrichmentException(ex);
+                }
+
+                if (this.options.RecordException)
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:248]: We should probably do a follow-up PR to add this to AspNet instrumentation too since they are very similar?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:248]: Need to if aspnet publishes an event with Exception in payload. If yes, we should add a follow up pR.

**Comment by eddynaka** [comment]
> @eddynaka Are you going to add a test for this?

i was thinking how i would do that.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:248]: I checked, it doesn't. My bad! 🤦

**Comment by CodeBlanch** [comment]
@eddynaka For the test...

https://github.com/open-telemetry/opentelemetry-dotnet/blob/41cd4f5d9c611b3d15187016bf1a70993464bc4e/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs#L146-L152

I think all you need is for that callback to throw an unhandled exception. You could have the test add something to request query string like "?throw=true" which you could detect off the context in the callback? Might work 😄

**Comment by eddynaka** [comment]
> @eddynaka For the test...
> 
> https://github.com/open-telemetry/opentelemetry-dotnet/blob/41cd4f5d9c611b3d15187016bf1a70993464bc4e/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs#L146-L152
> 
> I think all you need is for that callback to throw an unhandled exception. You could have the test add something to request query string like "?throw=true" which you could detect off the context in the callback? Might work 😄

Thanks for the guidance! just pushed one change testing the exception.
Let me know what do you think!

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs [hunk 6]
```diff
@@ -106,7 +107,7 @@ public IncomingRequestsCollectionsIsAccordingToTheSpecTests(WebApplicationFactor
 
             if (statusCode == 503)
             {
-                Assert.Equal(Status.Error, activity.GetStatus());
+                Assert.Equal(Status.Error.StatusCode, activity.GetStatus().StatusCode);
```
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs:110]: please add a test for this.options.RecordException as well. and i think this is good.

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs [hunk 7]
```diff
@@ -53,5 +53,10 @@ public class AspNetCoreInstrumentationOptions
         /// The type of this object depends on the event, which is given by the above parameter.</para>
         /// </remarks>
         public Action<Activity, string, object> Enrich { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the exception will be recorded or not.
+        /// </summary>
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:59]: will be recorded as `ActivityEvent` or not.
Add link to spec: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/exceptions.md

Comment (eddynaka) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:59]: done!

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs (hunk 6 above)
Comment (eddynaka) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/IncomingRequestsCollectionsIsAccordingToTheSpecTests.cs:113]: done!

**Review by cijothomas — APPROVED** [review]
